transform_assumption.ml

```diff
@@ -47,17 +47,64 @@ let insert_checkpoints (func:afunction) =
            { label = version.label;
              instrs = (|?) baseline instrs;
              annotations = None } ] }
+module LabelTarget = struct
+  type t = label position
+  let compare = Pervasives.compare
+end
 
-(* Removes all empty checkpoints. Can be applied as a final cleanup
- * when there are no osr-in points (because the osr instruction also
- * serves as a label for osr-entry). *)
-let remove_empty_osr : transform_instructions = fun input ->
-  let transform pc =
-    match[@warning "-4"] input.instrs.(pc) with
-    | Osr {cond} when cond = [] -> Remove 1
-    | _ -> Unchanged
+module TargetSet = Set.Make(LabelTarget)
+
+(* Removes all empty checkpoints with no incoming references *)
+let remove_empty_osr : opt_prog = fun prog ->
+  (* Collect the targets (including the extra frames continuations) of all non empty osrs *)
+  let used =
```

Comment:
At least the definition of `add` is directly above its use. That I can live with.